### PR TITLE
zfs: fix evaluation and build on aarch64 systems

### PR DESCRIPTION
**Pull request by Mic92**
This is blocking: https://github.com/nix-community/nixos-images/pull/115

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mic92** [comment]
As of now latestCompatibleLinuxPackages points to broken version on aarch64-linux.

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 1]
```diff
@@ -654,6 +668,21 @@ in
       services.udev.packages = [ cfgZfs.package ]; # to hook zvol naming, etc.
       systemd.packages = [ cfgZfs.package ];
 
+      # Export kernel_neon_* symbols again.
+      # This change is necessary until ZFS figures out a solution
+      # with upstream or in their build system to fill the gap for
+      # this symbol.
+      # In the meantime, we restore what was once a working piece of code
+      # in the kernel.
+      boot.kernelPatches = lib.optional (cfgZfs.removeLinuxDRM && pkgs.stdenv.hostPlatform.system == "aarch64-linux") {
```
Comment (lopsided98) [inline nixos/modules/tasks/filesystems/zfs.nix:677]: I think this should only apply the patch if the kernel is at least 6.2

**Review by lopsided98 — CHANGES_REQUESTED** [review]
I tried to test this PR with the following config, which I think is fairly standard.
```
boot.kernelPackages = config.boot.zfs.package.latestCompatibleLinuxPackages;
boot.zfs.removeLinuxDRM = true;
```

This still results in using 6.1, but also attempts to apply the patch which of course fails. This PR overrides the zfs package with `removeLinuxDRM` before it is passed to `extraModulePackages`, but this has no effect on the value of `boot.zfs.package` or `pkgs.zfs`, which are normally used to choose the proper kernel packages.

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 1 above)
Comment (Mic92) [reply nixos/modules/tasks/filesystems/zfs.nix:677]: Its not possible to check the kernel version here. This will lead to infinite recursion.

**Comment by Mic92** [comment]
Yes. This is a convoluted mess:

```
boot.kernelPackages = boot.kernelPackages = lib.mkDefault (pkgs.zfs.override {
  removeLinuxDRM = pkgs.hostPlatform.isAarch64;
}).latestCompatibleLinuxPackages;

boot.zfs.removeLinuxDRM = true;
```

If you have a better idea... I don't have one.

File: pkgs/os-specific/linux/zfs/unstable.nix [hunk 2]
```diff
@@ -13,19 +14,22 @@ callPackage ./generic.nix args {
...
+  else
+    linuxKernel.packages.linux_6_1;
 
   # this package should point to a version / git revision compatible with the latest kernel release
   # IMPORTANT: Always use a tagged release candidate or commits from the
   # zfs-<version>-staging branch, because this is tested by the OpenZFS
   # maintainers.
-  version = "2.1.12-staging-2023-04-18";
-  rev = "e25f9131d679692704c11dc0c1df6d4585b70c35";
+  version = "2.1.12";
```
Comment (RaitoBezarius) [inline pkgs/os-specific/linux/zfs/unstable.nix:30]: unstable should probably keep the staging stuff?

Comment (Mic92) [reply pkgs/os-specific/linux/zfs/unstable.nix:30]: Why?

Comment (RaitoBezarius) [reply pkgs/os-specific/linux/zfs/unstable.nix:30]: Hm, fair

Comment (RaitoBezarius) [reply pkgs/os-specific/linux/zfs/unstable.nix:30]: (usually, staging heads contains also performance/feature fixes, etc. But I agree that we want to focus on kernel updates)

**Comment by Mic92** [comment]
> I tried to test this PR with the following config, which I think is fairly standard.
> 
> ```
> boot.kernelPackages = config.boot.zfs.package.latestCompatibleLinuxPackages;
> boot.zfs.removeLinuxDRM = true;
> ```
> 
> This still results in using 6.1, but also attempts to apply the patch which of course fails. This PR overrides the zfs package with `removeLinuxDRM` before it is passed to `extraModulePackages`, but this has no effect on the value of `boot.zfs.package` or `pkgs.zfs`, which are normally used to choose the proper kernel packages.

I made this more explicit in the option description. Please have a look.

**Comment by Mic92** [comment]
friendly bump @RaitoBezarius @lopsided98

**Comment by RaitoBezarius** [comment]
I will take responsibility.

**Comment by howdoicomputer** [comment]
Are there plans to backport this to `23.05`?

**Comment by RaitoBezarius** [comment]
We are too close to 23.11 that I can take care of this. You are free to open a backport PR as long as it complies with the backport policies.

**Comment by truatpasteurdotfr** [comment]
hi, is this fixed for 23.11 ?

**Comment by Luflosi** [comment]
Yes, this is fixed in NixOS 23.11. You have to enable the `boot.zfs.removeLinuxDRM` option though: https://search.nixos.org/options?channel=23.11&show=boot.zfs.removeLinuxDRM

**Pull request merged**